debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java

```diff
@@ -117,38 +136,97 @@ private PostgresReplicationConnection(Configuration config,
+                    try (Statement stmt = pgConnection().createStatement(); ResultSet rs = stmt.executeQuery(selectPublication)) {
+                        if (rs.next()) {
+                            Long count = rs.getLong(1);
+                            if (count == 0L) {
+                                LOGGER.info("Creating new publication '{}' for plugin '{}'", publicationName, plugin);
+                                switch (publicationAutocreateMode) {
+                                    case ALL_TABLES:
+                                        // Publication doesn't exist, create it.
+                                        stmt.execute(String.format("CREATE PUBLICATION %s FOR ALL TABLES;", publicationName));
+                                        break;
+                                    case FILTERED:
+                                        try {
+                                            Set<TableId> tablesToCapture = determineCapturedTables(new RelationalSnapshotChangeEventSource.RelationalSnapshotContext(""));
+                                            String tableFilterString = tablesToCapture.stream().map(TableId::toDoubleQuotedString).collect(Collectors.joining(", "));
+                                            // Publication doesn't exist, create it but restrict to the tableFilter.
+                                            stmt.execute(String.format("CREATE PUBLICATION %s FOR TABLE %s;", publicationName, tableFilterString));
+                                        }
+                                        catch (Exception e) {
+                                            LOGGER.error("Unable to create publication for ");
+                                        }
+                                        break;
+                                }
+                            }
+                            else {
+                                LOGGER.trace(
+                                        "A logical publication named '{}' for plugin '{}' and database '{}' is already active on the server " +
+                                                "and will be used by the plugin",
+                                        publicationName, plugin, database());
+                            }
                         }
                     }
                 }
+                catch (SQLException e) {
+                    throw new JdbcConnectionException(e);
+                }
+
+                // This is what ties the publication definition to the replication stream
+                streamParams.put("proto_version", 1);
+                streamParams.put("publication_names", publicationName);
             }
-            catch (SQLException e) {
-                throw new JdbcConnectionException(e);
-            }
+        }
+    }
+
+    private Stream<TableId> toTableIds(Set<TableId> tableIds, Pattern pattern) {
```

Comment:
The name is a bit misleading, perhaps simply `filter()`?

Reply:
yeah this was copied from the `RelationalSnapshotChangeEventSource.determineCapturedTables(RelationalSnapshotContext)`